### PR TITLE
feat: analytics email service

### PR DESCRIPTION
**Pull request by GeronimoJohn**
# Description

Depends on #2873 

### Issue

Currently there are no ways for our creators to figure out if a users or visitors have interacted with their journeys.

[Set Up Email Service](https://3.basecamp.com/3105655/buckets/37663186/todos/7439501340)
[Create Email Analytics Templates](https://3.basecamp.com/3105655/buckets/37663186/todos/7439501236)

### Solution

We will set up an email service to notify our creators when a visitor interacts with their journey. An email will be queued if a user interacts with the Sign Up block, Text Response block, or clicks on Chat Buttons. This email will be sent after a 5-minute delay, which resets with each new interaction

### CODE REVIEW

File: apps/api-journeys/src/app/modules/event/email/email.service.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    await this.emailQueue.add(
+      'visitor-event',
+      {
+        journeyId,
+        visitorId
+      },
+      {
+        jobId,
+        delay: 2 * 60 * 1000, // delay for 2 minutes
+        removeOnComplete: true,
```
Comment (GeronimoJohn) [inline apps/api-journeys/src/app/modules/event/email/email.service.ts:30]: Note to self: remove on complete, need to have a check where jobId has already been sent, don't send an updated version of the job to the user

File: apps/api-journeys/src/app/modules/event/email/email.consumer.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+export interface ExtendedUserTeam extends UserTeam {
+  journeyNotifications: JourneyNotification[]
+}
+
+export interface ExtendedJourneys extends Journey {
+  userJourneys: ExtendedUserJourneys[]
+  team: Team & {
+    userTeams: ExtendedUserTeam[]
+  }
+}
```
Comment (Kneesal) [inline apps/api-journeys/src/app/modules/event/email/email.consumer.ts:42]: Use Prisma.UserJourneyGetPayload, Prisma.UserTeamGetPayload and Prisma.JourneyGetPayload for property types

e.g.
```
export type JourneyWithTeamAndUserJourney = Prisma.JourneyGetPayload<{
  include: {
    userJourneys: true
    team: true
    primaryImageBlock: true
  }
}>
```

File: apps/api-journeys/src/app/modules/event/email/email.consumer.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    await this.emailService.sendEmail({
+      to: data.user.email,
+      subject: `A visitor has interacted with your journey`,
+      text,
+      html
+    })
+  }
+
+  async sendEventsNotification(job: Job<EventsNotificationJob>): Promise<void> {
```
Comment (Kneesal) [inline apps/api-journeys/src/app/modules/event/email/email.consumer.ts:197]:
```suggestion
  async visitorEventEmails(job: Job<EventsNotificationJob>): Promise<void> {
```

File: apps/api-journeys/src/app/modules/event/email/email.consumer.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+          duration: true,
+          events: true
+        }
+      })
+    ])
+
+    return { journey, visitor }
+  }
+
+  async sendUserNotification(
```
Comment (Kneesal) [inline apps/api-journeys/src/app/modules/event/email/email.consumer.ts:143]:
```suggestion
  async visitorEventEmail(
```

File: apps/api-journeys/src/app/modules/event/email/email.consumer.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+          id: true,
+          createdAt: true,
+          duration: true,
+          events: true
+        }
+      })
+    ])
+
+    return { journey, visitor }
+  }
```
Comment (Kneesal) [inline apps/api-journeys/src/app/modules/event/email/email.consumer.ts:141]: lets move these to a libs folder nested in the directory as these don't pertain to sending email but are just helper functions , lets do DI to pass in services if needed

File: apps/api-journeys/src/app/modules/event/email/email.service.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+type VideoEvent = 'start' | 'play'
+
+@Injectable()
+export class EmailService {
+  constructor(
+    @InjectQueue('api-journeys-events-email')
+    private readonly emailQueue: Queue<EventsNotificationJob>
+  ) {}
+
+  async addVisitorEvent(
```
Comment (Kneesal) [inline apps/api-journeys/src/app/modules/event/email/email.service.ts:16]:
```suggestion
  async addVisitorEventEmailJob(
```

**Review by Kneesal — CHANGES_REQUESTED** [review]

File: apps/api-journeys/src/app/modules/event/email/email.consumer.ts (hunk 2 above)
Comment (GeronimoJohn) [reply apps/api-journeys/src/app/modules/event/email/email.consumer.ts:42]: This is cool

**Review by Kneesal — DISMISSED** [review]
lgtm

**Review by Kneesal — DISMISSED** [review]

**Review by Kneesal — DISMISSED** [review]

**Review by Kneesal — DISMISSED** [review]

**Review by Kneesal — APPROVED** [review]

**Pull request merged**